_posts/2025-11-17-new-dev-services-api.adoc

```diff
@@ -0,0 +1,196 @@
+- Provide an implementation of `io.quarkus.deployment.builditem.Startable` - extending `GenericContainer` and implementing `Startable` is a good pattern
+- Use the builders rather than directly constructing `DevServicesResultBuildItem`
+
+==== Get rid of static fields on the extension processor
+
+Extension authors should not rely on static variables for cross-instance communication. They should not assume that the invocation order of processors will be the same as the run order of applications.
+
+The https://quarkus.io/guides/writing-extensions#injection[extension writing guide] says “State should only be communicated between build steps by way of build items, even if the steps are on the same class.”
+However, almost every Dev Service implementation broke this rule, and used a static field to track previously-created services.
+
+A good heuristic when migrating to the new model is that all static fields should go away. For example, remove all fields like these ones:
+
+[source,java]
+----
+private static volatile RunningDevService devService;
+private static volatile MyDevServicesConfig capturedDevServicesConfiguration;
+private static volatile boolean first = true;
+
+----
+
+Deciding whether to re-use or replace a service is now handled centrally, based on a diff of the configuration.
+
+==== Get rid of shutdown logic
+
+Because service lifecycle is handled centrally, any shutdown listeners or other logic for stopping services should also be removed.
+
+=== Remove any references to `RunningDevService`
+
+Because the processor does not handle starting the service, it should never return a `RunningDevService`.
+
+==== Use the builder
+
+Instead of direct construction, use the new builder API. Choose `owned()` for services which are to be created,
+or `discovered()` to register externally-managed services which have been discovered.
+
+For example,
+
+[source,java]
+----
+    DevServicesResultBuildItem = DevServicesResultBuildItem.owned().name(MY_FEATURE_NAME)
+                    .serviceName(name)
+                    .serviceConfig(myConfig)
+                    .startable(() -> new MyContainer(
+                            myImageName,
+                            myConfig.port(),
+                            useSharedNetwork)
+                            .withEnv(myConfig.containerEnv())
+                     .configProvider(
+                            Map.of(someProp, s -> s.getConnectionInfo()))
+                    .build());
+----
+
+==== Eligibility for re-use
+
+How does the central lifecycle management decide whether a service can be re-used? This is based on 'sameness keys' (the config objects) passed in to the builder to use as the basis for the comparison.
+
+The key method is `.serviceConfig(myConfig)`. The current config is compared reflectively to the config of running services each restart.
+
+==== The `Startable`
+
+In order to support lazy starting, pass an implementation of `Startable` to the builder.
```

Comment:
It is obvious to us, but we can add here a note NOT to call `start` before passing the startable